src/main/java/com/google/research/bleth/utils/Queries.java

```diff
@@ -98,7 +98,18 @@
             }
         }
 
-        // Add last entity to result (if matching).
+        // Add remaining secondary entities to result (if matching).
+        while (secondaryEntityIterator.hasNext()) {
```

Comment:
Can you *prove* that `primaryKey.compareTo(secondaryKey) >= 0` here?
If not (i.e., `primaryKey.compareTo(secondaryKey) < 0`) - this loop will be infinite.